### DISCUSSION TITLE
Add configuration to disable ReactDevOverlay

### DISCUSSION THREAD

**Discussion by kylecesmat** [vercel/next.js#23970] [Ideas]
# Feature request
Developers should be able to opt-out of the `ReactDevOverlay` component that blocks the UI by providing configuration to `next.config.js`.

Something like:

```
// next.config.js
module.exports = {
    reactDevOverlay: false
}
```

This is similarly related to the `typescript.ignoreDevErrors` configuration that prevents the UI from breaking, but still provides developers with introspection in their console(s).

## Is your feature request related to a problem? Please describe.

No.

## Describe the solution you'd like

Add configuration here:
```
// next.config.js
module.exports = {
    reactDevOverlay: false
}
```

That prevents this component from rendering in development - https://github.com/vercel/next.js/blob/a0c683208abdce84934061f6efd2b90de52d99b9/packages/next/client/index.js#L160

## Describe alternatives you've considered

N/A

## Additional context

N/A

**Comment by timneutkens** [comment]
> Is your feature request related to a problem? Please describe.
> No.

So why do you want to disable the error overlay if there is no problem?

> That prevents this component from rendering in production

There is no error overlay in production as per that branch you linked to.

If you were referring to development errors that happen at runtime can be clicked away if needed. Compilation errors can't be clicked away given that they happen at compile time so executing the code will not work.

**Comment by Timer** [comment]
Also—see https://github.com/facebook/create-react-app/issues/6530#issuecomment-662914362 for the reasoning why this overlay cannot be disabled. We're going to improve its ergonomics with error boundaries, though!

**Comment by kylecesmat** [comment]
@timneutkens I apologize, I could have provided a better problem statement in the description.

The goal here would be to disable the development error banner that appears. You are right, it can be clicked away if needed, but that's just a developer annoyance when you are actively working and a known bug is throwing an error.

Ideally, **I'd like to be able to disable the rendering of this error boundary UI for development via a configuration value.**

My (rather unfortunate, but current) use case here is this:
- My nextjs application imports a library thats outputs an error in dev because it is flawed and incorrectly handles hostname parsing with `localhost`
- My team of 12 developers knows this library is an issue in dev, and the error "cannot read .lastIndexOf foo..." will cause this UI to display during every page transition while they are in development
- Ideally, I'd like if my team did not have to click the close button on every page transition just to close the modal of a bug they are ignoring intentionally in dev

yes, I know this is a contrived example, but ideally I'd like to have the control to enable/disable this developer tool and make my team's developer experience better while we work to fix the upstream issue.

To @Timer's point, it looks like this cannot be done currently? I was thinking it'd be as simple as adding a config value like this:

```
if (config.reactDevOverlay && process.env.NODE_ENV !== 'production') {
      const { ReactDevOverlay } = require('@next/react-dev-overlay/lib/client')
      return <ReactDevOverlay>{this.props.children}</ReactDevOverlay>
    }
```

**Comment by pacxiu** [comment]
We have similar usecase for such flag - Unhandled Runtime Error comes from chrome extension so it is displayed as `inpage.js`. It is not critical and we have to click it away on and on - becomes frustrating after a while.

**Comment by teeraphong** [comment]
this idea should have

**Comment by dontsave** [comment]
i want to endorse the idea of being able to disable or customize the error overlay such that it becomes less obtrusive. there are many cases in which a developer will be aware there are errors to solve, but will want to work on other aspects of the app regardless. the fullscreen overlay in these cases creates poor developer experience as the only way to remove them is to manually click to dismiss

**Comment by gabsn** [comment]
A `reactDevOverlay: false` would definitely be of great help here. Personally, I handle the exceptions with a custom component and this `<ReactDevOverlay />` keeps clashing with it...

**Comment by merrywhether** [comment]
Another reason this would be helpful: intentional errors.

We have our server making auth'ed API calls via `getInitialProps` (app is auth-only). If the API responds with 401 indicating an invalid auth token, we throw a custom error to get the `_error` page to render with a 401 status. Then we perform custom 401 logic client-side after app boot and either display custom 401 UI or redirect to login (depending on outcome of client-side actions). We could instead do more plumbing to send a flag from `gIP` to `_app` and have it dynamically render a `_401` page component or something, but re-using the `_error` path fits nicely into our logical map, where 401 is added to the switch statement that is already handling 404 and 500.

The non-configurable dev overlay means that we either have to:
1. Disable this code-path in dev, which feels really bad since this is auth-related code (some of the most important code in any app)
2. Develop the app via production builds so we can regularly see the effects of this logic

We'd be plenty happy finding unintentional errors in the server logs ourselves when they occur, and I'm not sure how you'd miss winding up with the 500 UI anyway (the dev overlay just seems like a nicety). Dan admits in the [referenced CRA comment](https://github.com/facebook/create-react-app/issues/6530#issuecomment-662914362) that intentional errors are not handled well by this (which is kind of ironic since suspense uses custom intentional errors itself). We'd `patch-package` the dev overlay out ourselves, but this app lives in a monorepo with other Next apps that do want the overlay behavior.

I guess another solution would be an easier official way to coalesce 401s into the _error page, but that's a much bigger ask.

**Comment by dougmorin** [comment]
This feature would be very useful.  Right now I'm troubleshooting adding my own error boundary and I'm fighting the dev overlay just to test what would happen in production.

**Comment by shanshanzhu** [comment]
I also wanna upvote this feature. Our use case is debugging an errorPage behavior.

**Reply by kostia1st** [reply]
This, exactly. I still fail to understand why Next.js team decided that we should NOT use `_error` and similar pages for controlled handling of expected errors. Developing and debugging this is such a pain currently 🤷‍♂️

**Comment by Marcosld** [comment]
For acceptance testing purposes with puppeteer we use these 2 functions to be able to test the unexpected error behaviour while hiding the overlay:
```typescript

    const disableReactErrorUI = async (page: PageApi) => {
        await page.addStyleTag({
            content: `
nextjs-portal {
    display: none;
}`,
        });
    };

    const bypassUnhandledRejections = async (page: PageApi) => {
        await page.evaluate(() => {
            window.addEventListener('error', (event) => {
                event.preventDefault();
            });
            window.addEventListener('unhandledrejection', (event) => {
                event.preventDefault();
            });
        });
    };
```
